eclipse-theia/theia
# Draft: Terminal Manager

#### What it does

Relevant to #12673

This is PR adds the package `terminal-manager`, which was built for our downstream IDE. It is heavily inspired by VSCode's split-panel terminal view + tree, but not an exact copy. The terminal manager is separate widget that supplements the existing terminals inside Theia but is not meant as a replacement (at the moment). The terminal-manager's purpose is to serve as a place to maintain and organize terminals to improve a user's way of working.

I'm hoping to hand off this feature to anyone that's interested as I will not have the bandwidth to bring this feature up to parity with VSCode. This hopefully will serve as a good jumping off point for some interesting terminal view improvements to Theia. Happy to answer any questions otherwise :)

# Description

![terminal-manager](https://github.com/eclipse-theia/theia/assets/62660714/8761692b-3cc4-48a4-9ca0-c97eba81fcc7)

**How to test**
- Build and run this branch
- Start Theia and run the command "Open Terminal Manager"
  - The terminal manager should appear in the bottom panel and look like this:

![image](https://github.com/eclipse-theia/theia/assets/62660714/5e823772-a86a-45ad-bb9e-320b8a3a8d95)

- On start you are given defaults of one page, one group, and one terminal, you can add additional pages by hovering over the treewidget and clicking the icons on the tree nodes
- Use the new keybinding `alt+q` to maximize and minimize the terminal panel. This keybinding is meant to serve as a convenient way of switching between the normal "editor" view and a fullscreen "terminal" view.  (you can also use the new toolbar item to maximize/minimize)

![image](https://github.com/eclipse-theia/theia/assets/62660714/961d9efd-9fe4-42b8-adb4-cdf3ec4ade58)

- Try adding groups, pages, terminals to get something like this:

![image](https://github.com/eclipse-theia/theia/assets/62660714/16e4a8ad-2041-481c-afc6-8487c8c00f24)

- Right click on tree nodes and try renaming nodes
- Click on different nodes, corresponding terminals should flash when their node is selected (and vice versa)
  - Pages should automatically update when their respective nodes are selected (and renamed)
- Try deleting groups, pages, terminals from the tree widget (you won't be allowed to delete the last page/group/terminal for its respective parent, does this behavior make sense?)

- Try showing and hiding the tree widget with this icon: 

![image](https://github.com/eclipse-theia/theia/assets/62660714/bdacecdf-b09a-4f2e-85bc-03b929ed4a38)

- If you want the tree widget to appear on the left/right, set this preference;

![image](https://github.com/eclipse-theia/theia/assets/62660714/f20e8202-b828-4bcd-8427-b5ebc2628891)

- The terminal manager's layout and state is stored in local storage
  - layouts should be stored with their corresponding workspace so a layout created in Workspace A will be saved whenever workspace A is opened, new workspaces/windows should start with a fresh terminal manager

#### How to test


#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
